gnat: init at 9

Lucus16 opened the pull request:
###### Motivation for this change

Gnat is the Ada compiler frontend in gcc. It used to be in nixpkgs based on gcc 4.5 but it broke when the cc wrappers were rewritten. I've been trying to get it to work with gcc 9 and the current cc wrapper scripts. I think I've ported all relevant parts, but for now, it seems the configure script is not convinced it has a gnat compiler available. I'm having a hard time figuring out why.

Currently, this PR needs a rebuild of the world because it needs some additions in `cc-wrapper/add-flags.sh`. An alternative option is to give gnat its own add-flags variant.

I don't have the time to finish this right now, so help is appreciated.

#### TODO

- [ ] Fix build non-determinism, see #9212.
- [ ] See if the `add-flags.sh` change can be avoided so a world rebuild isn't necessary.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Lucus16 committed:
gnat: init at 9.3.0

yorickvP commented:
cc @viric @vcunat @shlevy @matthewbauer
was disabled in https://github.com/NixOS/nixpkgs/commit/6ac7b19c978e951c124c5ea434c94f95f593888e
Also 
https://www.reddit.com/r/NixOS/comments/8eiw2s/id_like_to_get_ada_compiler_gnat_available_and/

yorickvP commented:
@ofborg build gnat

vcunat commented:
The configure script fails in this part
```autoconf
# Test for GNAT.
# We require the gnatbind & gnatmake programs, as well as a compiler driver
# that understands Ada.  We use the user's CC setting, already found, and
# possibly add $1 to the command-line parameters.
#
# Sets the shell variable have_gnat to yes or no as appropriate, and
# substitutes GNATBIND and GNATMAKE.
AC_DEFUN([ACX_PROG_GNAT],
[AC_REQUIRE([AC_CHECK_TOOL_PREFIX])
AC_REQUIRE([AC_PROG_CC])
AC_CHECK_TOOL(GNATBIND, gnatbind, no)
AC_CHECK_TOOL(GNATMAKE, gnatmake, no)
AC_CACHE_CHECK([whether compiler driver understands Ada],
		 acx_cv_cc_gcc_supports_ada,
[cat >conftest.adb <<EOF
procedure conftest is begin null; end conftest;
EOF
acx_cv_cc_gcc_supports_ada=no
# There is a bug in old released versions of GCC which causes the
# driver to exit successfully when the appropriate language module
# has not been installed.  This is fixed in 2.95.4, 3.0.2, and 3.1.
# Therefore we must check for the error message as well as an
# unsuccessful exit.
# Other compilers, like HP Tru64 UNIX cc, exit successfully when
# given a .adb file, but produce no object file.  So we must check
# if an object file was really produced to guard against this.
errors=`(${CC} $1[]m4_ifval([$1], [ ])-c conftest.adb) 2>&1 || echo failure`
if test x"$errors" = x && test -f conftest.$ac_objext; then
  acx_cv_cc_gcc_supports_ada=yes
fi
rm -f conftest.*])

if test "x$GNATBIND" != xno && test "x$GNATMAKE" != xno && test x$acx_cv_cc_gcc_supports_ada != xno; then
  have_gnat=yes
else
  have_gnat=no
fi
])
```

Apparently it desires that `$CC` can compile ADA and that doesn't hold?  (the gnatbind and gnatmake checks do succeed)  I don't know... I tried the `gnatboot` you added and `cc` didn't really work there, so perhaps that's the issue here?
```
$ cc -c hello.adb
gcc: error trying to exec 'gnat1': execvp: No such file or directory
```

Lucus16 commented:
`cc` is not from `gnatboot`, but rather from `stdenv`. I don't know about `${CC}`. I couldn't figure out how this worked before.

vcunat commented:
Oh, the command above actually works if I add `${gnatboot.cc}/bin` into `$PATH`, but the output from `gnat` build failure remains the same as far as I can see.

Lucus16 commented:
I just checked, it seems that originally, `CC=gcc` and both `gcc` and `cc` both point into the `gnatboot` package. We'll need to figure out some way to have `gnatboot` override `stdenv`. I'm not sure if the `stdenv` C compiler is still needed, but some C will need to be compiled.

Lucus16 commented:
`NIX_CC` still pointed to `stdenv` though.

vcunat commented:
Yes, I managed to fish the output from the test command:
```
gcc: error: conftest.adb: Ada compiler not installed on this system
```
I expect the build probably wouldn't work just with `gnatboot` itself.

Lucus16 commented:
I checked with `stdenvNoCC`, it doesn't work.

vcunat commented:
That's not what I meant.  I got
> checking for clearerr_unlocked... configure: error: C++ compiler missing or inoperational

Lucus16 commented:
By placing `gnatboot` in front of the `PATH` in the `preConfigure` phase, we get a lot further.

Lucus16 commented:
```
a-except.adb:860:42: missing argument for parameter "S" in call to "Image_Integer" declared at s-imgint.ads:39
make[3]: *** [../../gcc-9.1.0/gcc/ada/gcc-interface/Make-lang.in:1033: ada/libgnat/a-except.o] Error 1
```
Looks like an Ada error in the source, possibly because `gnatboot` is version 4.1 which might be too old to compile version 9.1.

vcunat commented:
Have you looked what bootstrap sequence is used elsewhere?

vcunat commented:
Applying the same approach to gnat6 leads even further &ndash; several minutes of compilation:
```
/nix/store/15vi6br717ndj6vd1cwqhc9pgvk2i5hy-binutils-2.31.1/bin/ld: cannot find crt1.o: No such file or directo
ry
/nix/store/15vi6br717ndj6vd1cwqhc9pgvk2i5hy-binutils-2.31.1/bin/ld: cannot find crti.o: No such file or directo
ry
collect2: error: ld returned 1 exit status
gnatlink: error when calling /build/build/gcc/xg++
make[3]: *** [../gcc-interface/Makefile:2626: common-tools] Error 4
make[3]: Leaving directory '/build/build/gcc/ada/tools'
make[2]: *** [Makefile:197: gnattools-native] Error 2
make[2]: Leaving directory '/build/build/gnattools'
make[1]: *** [Makefile:11308: all-gnattools] Error 2
```
Commit 8bd9463ae.

Lucus16 commented:
Looks like this still needs to be ported https://github.com/NixOS/nixpkgs/blob/b249f7f1a81765321d65c1f7861598ee4a5020a6/pkgs/development/compilers/gcc/gnat-cflags.patch

Lucus16 commented:
I ported it and I'm currently investigating other errors. Will push later today.

Lucus16 commented:
Just managed to build gnat6 using gcc6, but I will try to make it work using gcc7 for a bit.

vcunat commented:
I assumed that if the gentoo gnatboot could be used to build _some_ gnat version that's not too old, that one should hopefully be sufficiently new to build gnat9 and hopefully even some future versions.  (I tried 6 first for greater ease of modifications ATM due to gcc6 not being in stdenv bootstrap.)

Lucus16 commented:
`gnat6` works now, I'm currently building `gnat9`.

Lucus16 commented:
I managed to build `gnat9` using `gnat6`, but I cannot seem to build `ghdl` yet, so I suspect an issue in the wrapper.

Lucus16 commented:
So I just noticed the old gnat was broken in a sense: It sets the gnat cflags (for which the world rebuilding add flags patch is needed) to `-aI$basePath/adainclude -aO$basePath/adalib`, but `$basePath` was empty, so it turned into `-aI/adainclude -aO/adalib`, which are paths outside the nix store, so they got removed. Apparently they weren't necessary? I apparently fixed those paths and now it complains about `gcc` receiving `-aI` and `-aO` flags which it doesn't know what to do with. I'm going to try and see what happens if I drop all the gnat cflags stuff.

Lucus16 committed:
ghdl: init at 0.37

Lucus16 commented:
Turns out `basePath` wasn't empty for the wrapper around gnatboot and it is indeed necessary there to build gnat. It was empty in the wrapper around gnat and when it isn't as in this PR, it seems to get erroneously passed to the `gcc` in gnat, which can't handle them.

Lucus16 commented:
This is pretty annoying, it seems that on some runs, `gnat6` builds and on some runs, it fails with:
```
In file included from ../../gcc-6.5.0/gcc/ada/exit.c:42:0:
../../gcc-6.5.0/gcc/system.h:540:20: error: conflicting declaration of C function 'const char* strsignal(int)'
 extern const char *strsignal (int);
                    ^~~~~~~~~
In file included from /nix/store/x9hfv02kk0dmbvgza27cp4054pdcbpzc-gcc-7.4.0/include/c++/7.4.0/cstring:42:0,
                 from ../../gcc-6.5.0/gcc/system.h:235,
                 from ../../gcc-6.5.0/gcc/ada/exit.c:42:
/nix/store/z0wwgwjkzk5jvc00wiq21m68rmb5k08i-glibc-2.27-dev/include/string.h:446:14: note: previous declaration 'char* strsignal(int)'
 extern char *strsignal (int __sig) __THROW;
              ^~~~~~~~~
In file included from ../../gcc-6.5.0/gcc/system.h:685:0,
                 from ../../gcc-6.5.0/gcc/ada/exit.c:42:
../../gcc-6.5.0/gcc/../include/libiberty.h:112:14: error: ambiguating new declaration of 'char* basename(const char*)'
 extern char *basename (const char *) ATTRIBUTE_RETURNS_NONNULL ATTRIBUTE_NONNULL(1);
              ^~~~~~~~
In file included from /nix/store/x9hfv02kk0dmbvgza27cp4054pdcbpzc-gcc-7.4.0/include/c++/7.4.0/cstring:42:0,
                 from ../../gcc-6.5.0/gcc/system.h:235,
                 from ../../gcc-6.5.0/gcc/ada/exit.c:42:
/nix/store/z0wwgwjkzk5jvc00wiq21m68rmb5k08i-glibc-2.27-dev/include/string.h:483:26: note: old declaration 'const char* basename(const char*)'
 extern "C++" const char *basename (const char *__filename)
                          ^~~~~~~~
make[3]: *** [Makefile:1086: ada/exit.o] Error 1
```

Lucus16 commented:
My best guess is that the build order of certain components varies and some `include` is resolved differently depending on whether some component is built yet.

Lucus16 commented:
This seems to be related to #9212.

Lucus16 commented:
I've used the resulting GHDL successfully, but I don't think the intermittent build failure issue is resolved.

vcunat reviewed (approved):
Right.  I personally think a package that builds sometimes is more useful than no package :-)  given that a bad build should not succeed in this case AFAIK.  The code seems very safe wrt. risks to other packages.

Perhaps you should clean or explain the disabled parts of `pkgs/development/compilers/ghdl/default.nix`

Lucus16 commented:
Currently trying to get `ghdl-llvm` to build, I only got `ghdl-mcode` working before.

vcunat commented:
> Unhandled version llvm 7.1.0

vcunat commented:
It [accepts](https://github.com/ghdl/ghdl/blob/v0.36/configure#L252) versions starting with 7.0 (among others) but didn't expect our 7.1.0, though [it sounds relatively compatible](http://lists.llvm.org/pipermail/llvm-announce/2019-May/000083.html).  Still, if I patch that check, it does fail to build.

Lucus16 commented:
I reached the same conclusion. `cannot find crt1.o` and the warnings above it probably indicate that the wrong flags are passed, but I don't have enough understanding to figure out who's passing what flags and where they should come from.

Lucus16 commented:
Looking at the output with `NIX_DEBUG = 9;`, `ld` gets called with the `crt1.o` and `crti.o` arguments without a leading path or flag, I don't know if that's normal.

vcunat commented:
The same result is with `llvm_6`, so I expect version won't make a difference.

Lucus16 commented:
Earlier in the same build, `ld` gets passed full paths to `crt1.o` and `crti.o` and doesn't fail.

Lucus16 commented:
I've repinned this PR and fixed `ghdl-llvm` by setting `LIBRARY_PATH`. If Nix has a better solution than explicitly setting `LIBRARY_PATH`, let me know, otherwise, this PR is ready to be merged in my opinion.

Lucus16 commented:
@vcunat it would be nice if we can get this merged before the feature freeze in two days. Do you need me to do anything else? Please let me know.

vcunat commented:
I'll test the build locally and merge (today), but I'm afraid we're probably a week or two too late for the 19.09 fork-off, as it will be done from the master branch.  Still, I haven't kept a close look at the current plans, e.g. gcc8 default still isn't merged and "was planned" (originally).

vcunat commented:
Hmm, it still does not build for me: [dhx6br4f0m3kjnlvmc59ph7cpg54h3sv-ghdl-llvm-0.36.log](https://github.com/NixOS/nixpkgs/files/3580459/dhx6br4f0m3kjnlvmc59ph7cpg54h3sv-ghdl-llvm-0.36.log) &ndash; perhaps we'll disable that variant for now?

Lucus16 commented:
Looks like this is an instance of the intermittent build failure. It worked for me with 4 and with 8 cores. It might work if you try a few more times with `--cores 1`. Maybe we should just set `enableParallelBuilding = false;` on ghdl and gnat?

flokli commented:
If one derivation is flaky when building in parallel, setting `enableParallelBuilding = false` would be the right way.

Can you do that and rebase on top of latest staging?

As `langAda` doesn't default to true, and it mostly builds too, it'd probably be beneficial to have this in nixpkgs.

flokli commented:
Btw, I'm not so sure about the state of that `gnatboot` distfile on the gentoo servers.

If we need to use binaries, what about something like that as bootstrap?

```nix
{ stdenv
, fetchurl
, autoPatchelfHook
, ncurses5 }:

stdenv.mkDerivation {
  pname = "gnats-bootstrap";
  version = "2014";

  buildInputs = [
    ncurses5
  ];

  nativeBuildInputs = [
    autoPatchelfHook
  ];

  src = fetchurl {
    name = "gnat-gpl-2014-x86_64-linux-bin.tar.gz";
    url = "https://community.download.adacore.com/v1/6eb6eef6bb897e4c743a519bfebe0b1d6fc409c6?filename=gnat-gpl-2014-x86_64-linux-bin.tar.gz";
    sha256 = "0cxpxbx3dnq6fsy4k3qil377znv8k13nn3xgfha9jmpm9p4shqw0";
  };

  buildPhase = ":";

  installPhase = ''
    mkdir -p $out
    cp -R . $out

    rm $out/bin/gps_exe
    rm -R $out/lib/python2.7/ $out/share/gdb-7.7/python-2.7.3 $out/lib/gtk-3.0 $out/lib/gps/gdk-pixbuf-2.0 $out/lib/gps $out/lib/libpyglib-gi-2.0-python.so.0.0.0
  '';
}
```

There are also binaries for a lot of architectures, more than on the gentoo mirrors.

petabyteboy commented:
I've rebased this and was able to build a coreboot toolchain with Ada support, which is required for building coreboot with native graphics initialization for some boards.
Thanks for the work! My rebased version is available on the feature/gnat branch on my fork.

sylv-io commented:
@Lucus16 first of all, thanks for this PR.
I also [rebased](https://github.com/sylv-io/nixpkgs/tree/gnat) it and was able to build the coreboot toolchain with [this environment](https://github.com/sylv-io/nix-config/blob/T580/shells/coreboot.nix). I would love to see this PR merged. Do you have any plans to continue your work ?

Lucus16 commented:
It's been a while since I've worked on this. I'll see if I can rebase on staging and build a few times to see if the intermittent build failure is still present.

Lucus16 commented:
I rebased and squashed this. I built `gnat9` and `ghdl-mcode` once and `gnat6` a few times and did not encounter failures. I've marked `ghdl-llvm` as broken for now. I think this is ready to be merged.

@sylv-io I'd appreciate it if you could take over and find someone to merge this. I don't have much energy left to spend on this. I'll be around for questions.

sylv-io commented:
> I rebased and squashed this. I built `gnat9` and `ghdl-mcode` once and `gnat6` a few times and did not encounter failures. I've marked `ghdl-llvm` as broken for now. I think this is ready to be merged.
> 
> @sylv-io I'd appreciate it if you could take over and find someone to merge this. I don't have much energy left to spend on this. I'll be around for questions.

great thx! it's understandable, that your tired of this.
I will do my best to find some folks, but I'm quite new here.

vcunat reviewed (approved):
I re-checked the diff, builds on x86_64 (once), building and running a trivial hello-world program.  Even `ghdl-llvm` built for me :shrug: _On i686 all fails in `gnat6` step already, but that doesn't seem interesting._

In any case I believe this should get merged in some form; to me it seems OK as-is.

petabyteboy commented:
I just did a bunch of small changes:
- update ghdl to 0.37
- use pname
- quote homepage urls
- use https for homepage urls
- marked ghdl-llvm as unbroken
